File: backend/src/InsuraTech.Application/Policies/Commands/CreatePolicy/Strategies/CreateVehiclePolicyStrategy.cs
using InsuraTech.Domain.Policies;
using InsuraTech.Domain.Policies.ValueObjects;

namespace InsuraTech.Application.Policies.Commands.CreatePolicy.Strategies;

public sealed class CreateVehiclePolicyStrategy : ICreatePolicyStrategy
{
    public PolicyType Type => PolicyType.Vehicle;

    public bool CanHandle(CreatePolicyCommand request) =>
        !string.IsNullOrWhiteSpace(request.VehiclePlanId)
        && request.VehicleCommercialValue.HasValue
        && request.VehicleYear.HasValue
        && !string.IsNullOrWhiteSpace(request.VehicleBrand);

    public Task<Policy> CreateAsync(CreatePolicyCommand request, PolicyNumber policyNumber, InsuredPerson insured, CancellationToken cancellationToken)
    {
        var coverage = CoveragePeriod.Create(request.CoverageStartDate, request.CoverageEndDate);
        var policy   = Policy.CreateVehiclePolicy(
            policyNumber, insured, coverage,
            request.VehiclePlanId!,
            request.VehicleCommercialValue!.Value,
            request.VehicleYear!.Value,
            request.VehicleBrand!,
            DateOnly.FromDateTime(DateTime.UtcNow));

        return Task.FromResult(policy);
    }
}

File: backend_v2/src/main/java/com/insuratech/application/policies/strategies/CreateVehiclePolicyStrategy.java
// Spec: SPEC-010 — vehicle-quotation
// Origen: CreateVehiclePolicyStrategy.cs → CreateVehiclePolicyStrategy.java
package com.insuratech.application.policies.strategies;

import com.insuratech.application.policies.commands.CreatePolicyCommand;
import com.insuratech.domain.policies.Policy;
import com.insuratech.domain.policies.PolicyType;
import com.insuratech.domain.policies.vehicle.VehiclePlanCatalog;
import com.insuratech.domain.policies.vehicle.VehiclePlanSelection;
import com.insuratech.domain.policies.vehicle.VehicleQuotation;
import com.insuratech.domain.policies.vo.CoveragePeriod;
import com.insuratech.domain.policies.vo.InsuredPerson;
import com.insuratech.domain.policies.vo.PolicyNumber;
import org.springframework.stereotype.Component;

@Component
public class CreateVehiclePolicyStrategy implements PolicyCreationStrategy {

    @Override
    public PolicyType type() { return PolicyType.VEHICLE; }

    @Override
    public boolean canHandle(CreatePolicyCommand cmd) {
        return cmd.vehiclePlanId() != null && !cmd.vehiclePlanId().isBlank();
    }

    @Override
    public Policy create(CreatePolicyCommand cmd, PolicyNumber policyNumber, InsuredPerson insured) {
        var plan = VehiclePlanCatalog.findById(cmd.vehiclePlanId())
            .orElseThrow(() -> new IllegalArgumentException("Vehicle plan not found: " + cmd.vehiclePlanId()));
        var quotation = new VehicleQuotation(
            cmd.vehicleBrand(), cmd.vehicleModel(), cmd.vehicleYear(), cmd.vehicleCommercialValue());
        var selection = new VehiclePlanSelection(plan, quotation);
        var period = CoveragePeriod.of(cmd.coverageStartDate(), cmd.coverageEndDate());
        return Policy.createVehiclePolicy(policyNumber, insured, period, selection, cmd.agentId(), cmd.cityId());
    }
}
